arch/eventloop_lws_http.c

```diff
@@ -0,0 +1,259 @@
+    if(!el)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    /* Check the state */
+    if(cm->eventSource.state != UA_EVENTSOURCESTATE_STOPPED) {
+        UA_LOG_ERROR(el->logger, UA_LOGCATEGORY_NETWORK,
+                     "HTTP\t| To start the ConnectionManager, it has to be "
+                     "registered in an EventLoop and not started yet");
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    cm->eventSource.state = UA_EVENTSOURCESTATE_STARTED;
+    return UA_STATUSCODE_GOOD;
+}
+
+static void
+HTTP_eventSourceStop(UA_ConnectionManager *cm) {
+    if(cm->eventSource.state == UA_EVENTSOURCESTATE_STOPPING ||
+       cm->eventSource.state == UA_EVENTSOURCESTATE_STOPPED)
+        return;
+
+    UA_LOG_INFO(cm->eventSource.eventLoop->logger, UA_LOGCATEGORY_NETWORK,
+                "HTTP\t| Shutting down the ConnectionManager");
+
+    cm->eventSource.state = UA_EVENTSOURCESTATE_STOPPING;
+
+    /* Stopping and no connection open? */
+    HTTPConnectionManager *mcm = (HTTPConnectionManager*)cm;
+    if(mcm->cm.eventSource.state == UA_EVENTSOURCESTATE_STOPPING) {
+        mcm->cm.eventSource.state = UA_EVENTSOURCESTATE_STOPPED;
+        return;
+    }
+}
+
+static UA_StatusCode
+HTTP_eventSourceDelete(UA_ConnectionManager *cm) {
+    UA_String_clear(&cm->eventSource.name);
+    UA_free(cm);
+    return UA_STATUSCODE_GOOD;
+}
+
+/* The function is not needed for HTTP.
+ * Request can be sent directly with the sendWithConnection function. */
+static UA_StatusCode
+HTTP_openConnection(UA_ConnectionManager *cm, const UA_KeyValueMap *params,
+                    void *application, void *context,
+                    UA_ConnectionManager_connectionCallback connectionCallback) {
+    HTTPConnectionManager *hcm = (HTTPConnectionManager*)cm;
+
+    protocols[0].user = (void*)cm;
+
+    /* for the purpose of debugging */
+    //int logs = LLL_USER | LLL_ERR | LLL_WARN | LLL_NOTICE;
+    lws_set_log_level(0, NULL);
+
+    struct lws_context_creation_info info;
+    void *foreign_loops[1];
+
+    memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
+    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT |
+                   LWS_SERVER_OPTION_HTTP_HEADERS_SECURITY_BEST_PRACTICES_ENFORCE;
+    info.port = CONTEXT_PORT_NO_LISTEN; /* we do not run any server */
+    info.protocols = protocols;
+    info.connect_timeout_secs = 30;
+    info.event_lib_custom = &evlib_open62541; /* bind lws to our event loop */
+    foreign_loops[0] = hcm->cm.eventSource.eventLoop; /* pass in the event loop as the foreign loop object
+                                                      * we will bind to */
+    info.foreign_loops = foreign_loops;
+
+    hcm->context = lws_create_context(&info);
+    if (!hcm->context)
+        return UA_STATUSCODE_BADINTERNALERROR;
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+HTTP_sendWithConnection(UA_ConnectionManager *cm, uintptr_t connectionId,
```

Comment:
HTTP is synchronous.
So you have to wait for the previous request to receive a response before sending the next request.
Ensure there can be no misuse.

So either
- block sending requests while still waiting for a response
- Add the next requests to a list and send them out automatically later on.